src/device/socket/vsock.rs

```diff
@@ -222,62 +271,86 @@ impl<H: Hal, T: Transport> VirtIOSocket<H, T> {
         self.send_packet_to_tx_queue(&header, buffer)
     }
 
-    fn check_peer_buffer_is_sufficient(&mut self, buffer_len: usize) -> Result {
-        if usize::try_from(self.connection_info()?.peer_free())
-            .map_err(|_| SocketError::InvalidNumber)?
-            >= buffer_len
-        {
+    fn check_peer_buffer_is_sufficient(
+        &mut self,
+        connection_info: &mut ConnectionInfo,
+        buffer_len: usize,
+    ) -> Result {
+        if connection_info.peer_free() as usize >= buffer_len {
             Ok(())
         } else {
-            // Update cached peer credit and try again.
-            self.request_credit()?;
-            if usize::try_from(self.connection_info()?.peer_free())
-                .map_err(|_| SocketError::InvalidNumber)?
-                >= buffer_len
-            {
-                Ok(())
-            } else {
-                Err(SocketError::InsufficientBufferSpaceInPeer.into())
+            // Request an update of the cached peer credit, if we haven't already done so, and tell
+            // the caller to try again later.
+            if !connection_info.credit_requested {
+                self.request_credit()?;
+                connection_info.credit_requested = true;
```

Comment:
So this means, if the guest calling send() fails with `SocketError::InsufficientBufferSpaceInPeer` error, then it needs to call `wait_for_recv()` first to get the credit updated, then it can call `send()` again, right?

Reply:
Yes. And it might need to handle receiving some other data in the meantime.

Reply:
This approach may be implicit for an API. Although I understand that you want to avoid blocking the send() function, it may not be immediately obvious to users that a failed send() requires a call to recv() before retrying. Additionally, the usage of credit_requested may be unclear. Would it be possible to do a polling just after request_credit() ?

Reply:
I see. So the idea is to separate the interaction with the RX and TX queues into separate functions at the driver level, correct?

nit: Regarding `credit_requested`, I think renaming it to `has_pending_credit_request` could make its purpose clearer.

Reply:
I guess that's the implication, yes. I think it's probably still fine for `poll_recv` to send non-data packets to the TX virtqueue (e.g. responding to a credit request, shutdown or connection attempt), as that shouldn't block in most cases (though we do still need to figure out what to do in case the TX virtqueue is full, for some of these cases).

I've renamed `credit_requested` as you suggested.